Mathlib/Topology/Bornology/Absorbs.lean

```diff
@@ -0,0 +1,245 @@
+    {s : Set E} (hs : 0 ∈ s) : Absorbs M₀ s 0 :=
+  eventually_of_forall fun _ ↦ zero_subset.2 <| zero_mem_smul_set hs
+
+section GroupWithZero
+
+variable {G₀ α : Type*} [GroupWithZero G₀] [Bornology G₀] [MulAction G₀ α] {s t u : Set α}
+
+@[simp]
+protected lemma univ : Absorbs G₀ univ s :=
+  (eventually_ne_cobounded 0).mono fun a ha ↦ by rw [smul_set_univ₀ ha]; apply subset_univ
+
+lemma _root_.Set.Finite.absorbs_sInter {S : Set (Set α)} (hS : S.Finite) :
+    Absorbs G₀ (⋂₀ S) t ↔ ∀ s ∈ S, Absorbs G₀ s t := by
+  simp only [Absorbs, ← hS.eventually_all]
+  refine eventually_congr <| (eventually_ne_cobounded 0).mono fun a ha ↦ ?_
+  simp only [← preimage_smul_inv₀ ha, preimage_sInter, subset_iInter_iff]
+
+protected lemma sInter {S : Set (Set α)} (hs : S.Finite) (ht : ∀ s ∈ S, Absorbs G₀ s t) :
+    Absorbs G₀ (⋂₀ S) t :=
+  hs.absorbs_sInter.2 ht
+
+@[simp]
+lemma _root_.absorbs_inter : Absorbs G₀ (s ∩ t) u ↔ Absorbs G₀ s u ∧ Absorbs G₀ t u := by
+  simpa using ((finite_singleton t).insert s).absorbs_sInter
+#align absorbs_inter absorbs_inter
+
+protected lemma inter (hs : Absorbs G₀ s u) (ht : Absorbs G₀ t u) : Absorbs G₀ (s ∩ t) u :=
+  absorbs_inter.2 ⟨hs, ht⟩
+#align absorbs.inter Absorbs.inter
+
+@[simp]
+lemma _root_.absorbs_iInter {ι : Sort*} [Finite ι] {s : ι → Set α} :
+    Absorbs G₀ (⋂ i, s i) t ↔ ∀ i, Absorbs G₀ (s i) t :=
+  (finite_range s).absorbs_sInter.trans forall_range_iff
+
+protected lemma iInter {ι : Sort*} [Finite ι] {s : ι → Set α} (h : ∀ i, Absorbs G₀ (s i) t) :
+    Absorbs G₀ (⋂ i, s i) t :=
+  absorbs_iInter.2 h
+
+lemma _root_.Set.Finite.absorbs_biInter {ι : Type*} {I : Set ι} (hI : I.Finite) {s : ι → Set α} :
+    Absorbs G₀ (⋂ i ∈ I, s i) t ↔ ∀ i ∈ I, Absorbs G₀ (s i) t := by
+  simpa only [sInter_image, ball_image_iff] using (hI.image s).absorbs_sInter
+
+protected lemma biInter {ι : Type*} {I : Set ι} (hI : I.Finite) {s : ι → Set α}
+    (h : ∀ i ∈ I, Absorbs G₀ (s i) t) : Absorbs G₀ (⋂ i ∈ I, s i) t :=
+  hI.absorbs_biInter.2 h
+
+@[simp]
+lemma _root_.absorbs_zero_iff [NeBot (cobounded G₀)] {E : Type*} [AddMonoid E]
+    [DistribMulAction G₀ E] {s : Set E} : Absorbs G₀ s 0 ↔ 0 ∈ s := by
+  refine ⟨fun h ↦ ?_, .zero⟩
+  rcases (h.and (eventually_ne_cobounded 0)).exists with ⟨c, hc, hc₀⟩
+  rcases hc rfl with ⟨x, hx, hx₀⟩
+  rwa [← smul_zero c⁻¹, ← hx₀, inv_smul_smul₀ hc₀]
+#align absorbs_zero_iff absorbs_zero_iff
+
+end GroupWithZero
+
+end Absorbs
+
+section AddGroup
+
+variable {M E : Type*} [Monoid M] [AddGroup E] [DistribMulAction M E] [Bornology M]
+
+@[simp]
+lemma absorbs_neg_neg {s t : Set E} : Absorbs M (-s) (-t) ↔ Absorbs M s t := by simp [Absorbs]
+
+alias ⟨Absorbs.of_neg_neg, Absorbs.neg_neg⟩ := absorbs_neg_neg
+#align absorbs.neg Absorbs.neg_neg
+
+lemma Absorbs.sub {s₁ s₂ t₁ t₂ : Set E} (h₁ : Absorbs M s₁ t₁) (h₂ : Absorbs M s₂ t₂) :
+    Absorbs M (s₁ - s₂) (t₁ - t₂) := by
+  simpa only [sub_eq_add_neg] using h₁.add h₂.neg_neg
+#align absorbs.sub Absorbs.sub
```

Comment:
Can you multiplicativise this?

Reply:
What assumptions do you have in mind? `MulDistribMulAction`? Something else? I assumed that we're going to use `Absorbs` only for (generalizations of) modules.

Reply:
I was thinking `MulDistribMulAction`. If you think that's nonsensical (because it would never be used in applications), I'm happy for you to say so in an implementation note.

However I suspect it might be useful for geometric group theory, where groups are usually multiplicative and we care about the growth of iterated products of sets (and there `Absorbs` means the same as "is contained in the subgroup closure" if I'm not wrong).

Reply:
I'm going to add an implementation note saying that we'll add `MulDistribMulAction` versions when someone needs them.